Docker Feature

Killerherts opened the pull request:
This will add the docker file needed as well as an entry file to be run on the containers start. I ended up needing to move setup.py so that ether setup.py or BingRewards.py could be run at the start of the container. This already has a cron entry setup for once every 8hours with a random delay. 

This breaks update.sh, I don't want to hard code the git clone into the script so you will not have an established git tree and the files will already exist. One option is to clone to temp and then move the  .git folder over but I have not gotten there yet. 
I can work on that in a few days. For now everything besides update.sh is working. 


This image has been push to killerherts/bing-rewards:DEV for now until merged with main branch

pironic commented:
fyi, the 
``` ## Container Installation
1. Open a terminal and use the following commands
2. docker pull killerherts/bing-rewards:latest
3.A docker run -t -d --name bing-rewards killerherts/bing-rewards:latest python setup.py -e <your_email> -p <password> 
3.B Alternatively you can pass your config volume directly in with: docker run -t -d -v <path-to-config.json>:/bing-rewards/BingRewards/config/config.json --name bing-rewards killerherts/bing-rewards:latest
4. To enter the container for maintance user docker exec -it bing-rewards /bin/bash container will be setup to run script once every 8 hours this can be modified using crontab -e 
 ```

renders as:

![image](https://user-images.githubusercontent.com/541154/184948007-9ed4df78-51d7-42ed-850c-2637245a3468.png)

very confusing. might need a small adjustment.

pironic commented on README.md:
```diff
@@ -39,9 +39,10 @@ Here's an example of running non-default arguments
-3.B Alternatively you can pass your config volume directly in with: docker run -t -d -v <path-to-config.json>:/bing-rewards/BingRewards/config/config.json --name bing-rewards killerherts/bing-rewards:latest
-4. To enter the container for maintance user docker exec -it bing-rewards /bin/bash container will be setup to run script once every 8 hours this can be modified using crontab -e 
+3. docker run -t -d --name bing-rewards killerherts/bing-rewards:latest python setup.py -e <your_email> -p <password>  
+Alternatively you can pass your config volume directly in with: docker run -t -d -v <path-to-config.json>:/bing-rewards/BingRewards/config/config.json --name bing-rewards killerherts/bing-rewards:latest
```
if this is to be merged then it should be the target repo, no?

jjjchens235 vs killerherts

Killerherts replied:
> if this is to be merged then it should be the target repo, no?
> 
> jjjchens235 vs killerherts

This would be on jj if he wants to maintain a docker image as well. But I have found the person maintaining the git repo doesn't also want the extra work of maintaining the docker hub image as well, so another maintainer will host it. 
I will bow to jj on this one, but I doubt he wants to maintain the docker image as well. If he does, we can change that.

pironic commented on Dockerfile:
```diff
@@ -1,27 +1,29 @@
+    /bing-rewards/entry.sh \
+    && chmod 0744 /bing-rewards/script.sh
+RUN chown -R nonroot /bing-rewards 
+RUN chmod u+s /usr/sbin/cron 
```
more than one way to peel a potato but this adds a bunch of docker layers. Just curious why so many disconnected RUN commands? Makes it easier to debug i'm sure but would it be worth it to merge with `&& \`

Killerherts replied:
Correct. This is for debugging one blob of runs is a pain to troubleshoot. I did a version with one giant run and the build time was marginally less. Someone with more knowledge than me might have a compelling reason to condense it if so we totally can.

pironic replied:
rtfm'ing: looks like if it doesn't add too much size then there isn't really a need to merge layers so much these days; https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#minimize-the-number-of-layers

As long as each layer only contains the essentials to the final, it's fine. good to know!

Killerherts replied:
I did some reading your correct that would be best practice.

Killerherts commented:
@pironic I made some changed should be better now 

Made some changes added a script to add some clearer information about cron scheduling as well 
Reminder the image will remain on DEV tag until merged.

@jjjchens235 do you have changes you would like to be made?

jjjchens235 commented:
Sorry, I haven't had a chance to review this yet, I will take a look when I can. Appreciate all the work on this.

pironic commented on compose.yaml:
```diff
@@ -0,0 +1,16 @@
+services:
+    bing-rewards:
+        container_name: bing-rewards
+        image: killerherts/bing-rewards
```
what will the future workflow for this be? 

@jjjchens235 do you have any intention to update a dockerhub image when changes are pushed to your master branch?

pironic replied:
could be an option to include a ` build .` instead of pulling from dockerhub. could also set up a githubaction to auto-build and pull from github's docker repo.

Killerherts replied:
I considered pulling from GitHub but the problem is if the repo is gone or changed to a certain extent, it could break the script and build. I prefer to have the images already have everything they need. 

Honestly, the easy opportunity here is to just wget the zip file and just merge the folders. I just have not had time to write up the shell script for it yet.

pironic replied:
if the repo is gone even a wget zip wouldn't be guaranteed. safest way is to just local build. the only case where that wont work is in a swarm/k8s enviro and that would be pulled/deployed by someone who would be able to identify the issue and fix it themselves anyway. encourage a local build and you dont have to be responsible for maintaining the image repo :)

ForceConstant replied:
Should automate building and pushing to github container registry when a new release is made. https://www.howtogeek.com/devops/how-to-set-up-automatic-builds-for-docker-images-on-github/  Basically have a docker-publish.yml that kicks off a github workflow when branch is updated.

Killerherts replied:
@ForceConstant this sounds like the best solution. I don't want to make more changes until this is merged with the dev branch and we decide whose maintaining the image as most other these things will build off those foundational decisions by jj.

Killerherts commented:
BingRewards.py has been tested for 2 days on windows, unraid and linux, setup.py is also working as expected.
Update.sh still needs to be change to be compatable with docker installation I would suggest we just move away from git.
Telegram and other intergrations have not been tested yet.

jjjchens235 commented:
# Issue 1: downloading chromedriver

Can the python code in DockerHub be updated to use the most current code in master?

The most recent change that removes `cdc_` string replacement logic is not included in this image, and it's causing me to unsuccessfully run the code.

After I rebuilt the docker image using the most updated code from master it worked for me.

On a related note, I don't know much about this area of auto updating the image, but it seems like you and ForceConstant are in agreement to go this route so we can plan for this:

`Should automate building and pushing to github container registry when a new release is made. https://www.howtogeek.com/devops/how-to-set-up-automatic-builds-for-docker-images-on-github/ Basically have a docker-publish.yml that kicks off a github workflow when branch is updated.`

# Issue 2: setup.py
The following `setup.py` instructions in the README don't work for me:
```sh
docker run -t -d --name bing-rewards killerherts/bing-rewards:latest python setup.py -e <your_email> -p
```
The above command does not prompt me for my password and when I go into the container, the config.json file is never created.

Neither does this command
```sh
docker run -t -d -v <ABSOLUTE-path-to-config.json>:/bing-rewards/BingRewards/config/config.json --name bing-rewards killerherts/bing-rewards:latest
```
The `config/` directory is not copied over.

# Issue 3: readme clarity
Can we please add some markdown highlighting to make the instructions more clear? 

Something like:

```md
## Container Installation
1. In terminal, run `docker pull killerherts/bing-rewards:latest`
2. Set-up the config with either option 1 or 2 
	1. Option 1, run setup.py again: `docker run -t -d --name bing-rewards killerherts/bing-rewards:latest python setup.py -e <your_email> -p <password>` 
	2. Option 2: Pass your config volume directly into the container: `docker run -t -d -v <absolute-path-to-config.json>:/bing-rewards/BingRewards/config/config.json --name bing-rewards killerherts/bing-rewards:latest`
3. To enter the container for maintenance `docker exec -it bing-rewards /bin/bash` 
Note: initially the container will be setup to run script once every 8 hours this can be modified using `docker exec -it bing-rewards /bin/bash crontab -e`
```

# Issue 4: number of commits
Can we please squash these commits into a more manageable size.

# Issue 5: building the docker file
When I try to build the DockerFile manually (instead of pulling it), I run into this error:
```sh
#11 307.8 + chmod 0644 /etc/cron.d/bing.cron
#11 307.9 + chmod +x /bing-rewards/update.sh
#11 307.9 + /bing-rewards/entry.sh
#11 307.9 /bin/sh: 1: /bing-rewards/entry.sh: Permission denied
------
executor failed running [/bin/sh -c set -ex     && apt-get update --no-install-recommends -y     && apt-get install --no-install-recommends -y         chromium        chromium-driver        git        nano        cron     && rm -rf /var/lib/apt/lists/*     && touch /var/log/cron.log     && adduser --system nonroot     && chmod 0644 /etc/cron.d/bing.cron     && chmod +x /bing-rewards/update.sh     && /bing-rewards/entry.sh     && chmod 0744 /bing-rewards/script.sh     && chown -R nonroot /bing-rewards     && chmod u+s /usr/sbin/cron]: exit code: 1
```

The fix was adding the following line to the DockerFile
```sh
	# this line specifically:
    && chmod +x /bing-rewards/entry.sh \
    && /bing-rewards/entry.sh \
```

Killerherts commented:
### Issue 1  
---
Honestly, I am not versed in this area and would need time to figure it out. Maybe forced is more versed in this and could save us sometime.

### Issue 2
---
 will work if we include the password in the run line. If you don't want to include your password in the run cmd you need the run the container with -it but the problem with that will be that you will need to keep the windows open to keep the container running. This isn't really solvable as its just how docker works from the run command if your in -t -d mode it will skip over any user prompts during run process but keeps container running in background. 
```docker run -t -d -v <ABSOLUTE-path-to-config.json>:/bing-rewards/BingRewards/config/config.json --name bing-rewards killerherts/bing-rewards:latest```
Can be fixed by not mounting the file but the directory ```<ABSOLUTE-path-to-config>:/bing-rewards/BingRewards/config>```

### Issue 5
---
Looks like fix is to remove the && at line 21. 20 and 21 are already the 2 lines you had to add to make it build. @jjjchens235 Do you want me to continue to host the image or should I update the readme and instructions to your username?

jjjchens235 commented:
Regarding issue 1, wouldn't you simply update/merge your repo with the new code in master?

Then you would rebuild the Docker image and then commit it to Dockerhub?


Makes sense for issue #2, I'll test it out later.

Killerherts commented:
###Current Status
---
- [ ] Auto-update
- [x] Added clarity for why you need to include your password and not wait for a prompt
- [x] Updated ReadMe to match jj suggestion
- [x] Squashed commits into one as this is all for the docker feature  
- [x] Removed && from docker file this was a typo

Killerherts commented:
I solved the issue with update/sh by cloning the repo during entry.sh only coping the .git directory to /bing-rewards then deleting the temp folder. Users this allows people to just use update.sh as intended. If you want me to add a cron entry for auto update, let me know I like to manage my update on my own. 

This solution avoids using docker hub integration due to it being a paid service.

jjjchens235 commented:
I have been trying to get the following to work:

Option 1, run setup.py again: `docker run -t -d --name bing-rewards killerherts/bing-rewards:latest python setup.py -e <your_email> -p <password>` You must include your password as there will be no user prompt with -t -d

What happens is that the command runs successfully, but the container immediately exits. I believe the issue is that this line in the Dockerfile is not being executed  `CMD ["tail","-f","/dev/null"]` because it's being overriden by the `python setup.py` command. 

This is the output of `docker logs <image>` using the image created from the above command. Note that the tail command is never executed:
```sh
Requirement already satisfied: pyasn1<0.5.0,>=0.4.6 in /usr/local/lib/python3.9/site-packages (from pyasn1-modules>=0.2.1->google-auth<3.0.0dev,>=1.19.0->google-api-python-client->-r /bing-rewards/BingRewards/requirements.txt (line 7)) (0.4.8)
Requirement already satisfied: h11<1,>=0.9.0 in /usr/local/lib/python3.9/site-packages (from wsproto>=0.14->trio-websocket~=0.9->selenium>=4.0.0->-r /bing-rewards/BingRewards/requirements.txt (line 3)) (0.13.0)
+ mkdir /bing-rewards/BingRewards/logs
mkdir: cannot create directory ‘/bing-rewards/BingRewards/logs’: File exists
+ python setup.py -e dunhate235@gmail.com -p 'Yurmom5957!'

Warning: User set the password in plain text. Use `-p` with no arguments next time for better security.

config.json created/updated successfully
```


In contrast, this is the output of `docker logs <image>` using the volume command. Note the tail command is executed:
```sh
Requirement already satisfied: pyasn1<0.5.0,>=0.4.6 in /usr/local/lib/python3.9/site-packages (from pyasn1-modules>=0.2.1->google-auth<3.0.0dev,>=1.19.0->google-api-python-client->-r /bing-rewards/BingRewards/requirements.txt (line 7)) (0.4.8)
Requirement already satisfied: h11<1,>=0.9.0 in /usr/local/lib/python3.9/site-packages (from wsproto>=0.14->trio-websocket~=0.9->selenium>=4.0.0->-r /bing-rewards/BingRewards/requirements.txt (line 3)) (0.13.0)
+ mkdir /bing-rewards/BingRewards/logs
mkdir: cannot create directory ‘/bing-rewards/BingRewards/logs’: File exists
+ tail -f /dev/null
```

On a side-note, both logs are indicating that `entry.sh` file is throwing an error and aborting prematurely:
`mkdir /bing-rewards/BingRewards/logs
mkdir: cannot create directory ‘/bing-rewards/BingRewards/logs’: File exists`

I think adding a -p flag to mkdir in entry.sh should fix this issue.

jjjchens235 commented:
> I solved the issue with update.sh by cloning the repo during entry.sh only coping the .git directory to /bing-rewards then deleting the temp folder. Only downside to this solution is it will think the files are up to date on when you run the container so will only update once the head changes on your still running container. The docker image would just have to be kept up to date manually by me. We could also add a fetch and reset --hard and that would solve the image being "updated" at build problem but breaks the container if you're working on a forked branch as it will overwrite all your local files with jj's
> 
> This allows people to just use update.sh as intended. If you want me to add a cron entry for auto update, let me know I like to manage my update on my own.
> 
> This solution avoids using docker hub integration due to it being a paid service.

I thought you said that `GitHub Container Registry` is a paid service for images over 500MB, but that `DockerHub` would be free?
It looks like GitHub Actions and DockerHub interface together as well, based on   [Github's docs](https://docs.github.com/en/actions/publishing-packages/publishing-docker-images#publishing-images-to-docker-hub).

Killerherts commented:
> > I solved the issue with update.sh by cloning the repo during entry.sh only coping the .git directory to /bing-rewards then deleting the temp folder. Only downside to this solution is it will think the files are up to date on when you run the container so will only update once the head changes on your still running container. The docker image would just have to be kept up to date manually by me. We could also add a fetch and reset --hard and that would solve the image being "updated" at build problem but breaks the container if you're working on a forked branch as it will overwrite all your local files with jj's
> > This allows people to just use update.sh as intended. If you want me to add a cron entry for auto update, let me know I like to manage my update on my own.
> > This solution avoids using docker hub integration due to it being a paid service.
> 
> I thought you said that `GitHub Container Registry` is a paid service for images over 500MB, but that `DockerHub` would be free? It looks like GitHub Actions and DockerHub interface together as well, based on [Github's docs](https://docs.github.com/en/actions/publishing-packages/publishing-docker-images#publishing-images-to-docker-hub).

docker charges for their version of it as well

Killerherts commented:
moved tail to entry to catch the mkdir error if it already exists

jjjchens235 commented:
Beautiful! This last commit fixes the `setup.py` issue, so I think we're good to go here!

Only two things I can think of is:
- a small blurb in the README container section about `update.sh`
- special request: can we include `vim` in the image? I really dislike nano.


What do you think?

jjjchens235 commented:
> > > I solved the issue with update.sh by cloning the repo during entry.sh only coping the .git directory to /bing-rewards then deleting the temp folder. Only downside to this solution is it will think the files are up to date on when you run the container so will only update once the head changes on your still running container. The docker image would just have to be kept up to date manually by me. We could also add a fetch and reset --hard and that would solve the image being "updated" at build problem but breaks the container if you're working on a forked branch as it will overwrite all your local files with jj's
> > > This allows people to just use update.sh as intended. If you want me to add a cron entry for auto update, let me know I like to manage my update on my own.
> > > This solution avoids using docker hub integration due to it being a paid service.
> > 
> > 
> > I thought you said that `GitHub Container Registry` is a paid service for images over 500MB, but that `DockerHub` would be free? It looks like GitHub Actions and DockerHub interface together as well, based on [Github's docs](https://docs.github.com/en/actions/publishing-packages/publishing-docker-images#publishing-images-to-docker-hub).
> 
> docker charges for their version of it as well essentially both charge $5 per month for the ability to integrate build on repo push.

Ok thanks for the clarification. So just to make sure we're on the same page, because both GitHub Registry and DockerHub both charge $5/mo for auto-repo updates, we will not use this feature.

Instead, you will continue to host the image in DockerHub so it can be pulled down, and the user can simply run `update.sh` to get any new code from the master repo. Do I understand everything correctly? 

I think this solution should work fine as I would imagine the image should require almost no updates going forward, and the user having to run `update.sh` is pretty painless.

You had said `Only downside to this solution is it will think the files are up to date on when you run the container so will only update once the head changes on your still running container.` This is the one part I don't follow.

Killerherts commented:
> Beautiful! This last commit fixes the `setup.py` issue, so I think we're good to go here!
> 
> Only two things I can think of is:
> 
> * a small blurb in the README container section about `update.sh`
> * special request: can we include `vim` in the image? I really dislike nano.
> 
> What do you think?

Oh man I am not a vim fan I don't wanna remember extra buttons lol but yes I can add it on. I can add something about update.sh but it's actually not part of my additions. Should I add a cron entry for it once a day?

jjjchens235 commented:
`So because of the way I copy the .git folder it will think what ever version you have at that time is what it just downloaded. It will only notice the head change if the container is already running.`

I'm still not following here. Could you give me an example of where this is an issue? I would imagine that if the container code was pulled with commit_n, and the master repo was actually at commit_n+1, updating with a git pull in the container would get the container to be at commit_n+1 as well.

Killerherts commented:
I would imagine that if the container code was pulled with commit_n, and the master repo was actually at commit_n+1, updating with a git pull in the container would get the container to be at commit_n+1 as well.

When the container is pulled only makes a .git tree that is already at whatever the master. So it thinks it's already up to date. Until the master changes when the container is running it won't pull because the .git folder is always a copy of the most current when you pull the image.

My update.sh (work in progress) rework should handle this allowing you to do a hard reset if you want.

jjjchens235 commented:
I see. The `.git` directory is being pulled from the remote repo directly, per this line in entry.sh:
```sh
git clone https://github.com/jjjchens235/bing-rewards.git temp
mv temp/.git  /bing-rewards/.git
```

But there's no guarantee that the code in the docker image is at parity.


I'm not a fan of having `git reset --hard` command in update.sh, i don't think we should be rewriting history especially if the user is not aware that it is happening.


Here are some other options I can think of:
### Option 1
The `.git/` should already be based off the code in the image, is this not possible?

### Option 2
Currently the bingrewards code in the container is based off of: `COPY /BingRewards  /bing-rewards/BingRewards`.
In the DockerFile, replace this code with a `git clone jjjchens235/bing-rewards` instead.

Killerherts commented:
> I see. The `.git` directory is being pulled from the remote repo directly, per this line in entry.sh:
> 
> ```shell
> git clone https://github.com/jjjchens235/bing-rewards.git temp
> mv temp/.git  /bing-rewards/.git
> ```
> 
> But there's no guarantee that the code in the docker image is at parity.
> 
> I'm not a fan of having `git reset --hard` command in update.sh, i don't think we should be rewriting history especially if the user is not aware that it is happening.
> 
> Here are some other options I can think of:
> 
> ### Option 1
> The `.git/` should already be based off the code in the image, is this not possible?
> 
> ### Option 2
> Currently the bingrewards code in the container is based off of: `COPY /BingRewards /bing-rewards/BingRewards`. In the DockerFile, replace this code with a `git clone jjjchens235/bing-rewards` instead.



> I see. The `.git` directory is being pulled from the remote repo directly, per this line in entry.sh:
> 
> ```shell
> git clone https://github.com/jjjchens235/bing-rewards.git temp
> mv temp/.git  /bing-rewards/.git
> ```
> 
> But there's no guarantee that the code in the docker image is at parity.
> 
> I'm not a fan of having `git reset --hard` command in update.sh, i don't think we should be rewriting history especially if the user is not aware that it is happening.
> 
> Here are some other options I can think of:
> 
> ### Option 1
> The `.git/` should already be based off the code in the image, is this not possible?
> 
> ### Option 2
> Currently the bingrewards code in the container is based off of: `COPY /BingRewards /bing-rewards/BingRewards`. In the DockerFile, replace this code with a `git clone jjjchens235/bing-rewards` instead.

Option 2 
Is bad practice with docker to do a git clone inside the build or pull as it would break the container in the future if the maintainer change or your branch of the code goes dead.

update.sh has never worked for me with this bot long before I added the docker feature. People don't even know how to use git.

I am not sure what your wanting as any solution is going to replace history here as there is not tree to begin with.

I don't see why allowing users to force replace files for an update if they desired. If documented in a readme I don't see an issue. 

Git clone during build just isn't an option it breaks stuff for people doing development.


But as it stands now update.sh doesn't work unless you did a git clone of the master it's always been this way.

jjjchens235 commented:
## Original goal
Ok, let's back up here real quick. The original intent of this discussion was to figure out how to have the docker container auto-update when there are changes to the master repo right?

## confusion on original intent
So I'm confused when you write the following because I thought being at parity via auto-update is the last piece of the puzzle here:
`It was never the intention that the files in the container would be at parity nor is it a goal for me. I ofen make changes that other may not want so I am giving people that option by not forcing file replacement on build.`

## auto-update methods
#### github container registry
Originally, we were going to go with Force's idea, which was to `automate building and pushing to GitHub container registry when a new release is made.`

#### update update.sh
However, this has a monthly cost, so you're suggesting we use `update.sh` instead to account for new changes to the repo.

You're saying that `update.sh` code needs to be updated because the `.git/` directory in the container does not represent the actual state of the container, it represents the remote repo. Therefore, running `update.sh` will not work because based off the .git directory, it thinks it's already at the most recent remote master change.

#### keep update.sh, update DockerFile
Wouldn't the simplest solution be to include the `.git/` directory that represents the actual state of the container?
Adding something like this to the DockerFile:
```
COPY /.git /bing-rewards/.git
```

#### Cloning
`This is bad practice to do a git clone inside the build or pull as it would break the container in the future if the maintainer change or your branch of the code goes dead.`

I mean that would be one disadvantage of cloning. 
My understanding is that in general, cloning is an acceptable method with pros/cons, [source1](https://github.com/telefonicaid/perseo-fe/issues/498#issue-818717966), [source2](https://dzone.com/articles/clone-code-into-containers-how).

For our purposes, it goes back to the discussion about being at parity, i.e Do we want the docker container code to reflect the master branch?


## Further replies
1. `I am not sure what your wanting as any solution is going to replace history here as there is not tree to begin with inside the container. `
2. `I think giving the end user the option to force their build to be current is a reasonable solution as if they don't like what just happened all they have to do is compose down the container and pull it again `

Update.sh is already being used by people outside of a docker environment, where this is NOT true.

Killerherts commented:
>Update.sh is already being used by people outside of a docker environment, where this is NOT true.

For those who simply download and copy the zip (this is how i originally installed I did not use git at the time) the update script doesn't work as they don't have a tree to check out. I was just saying a flag to allow someone to force recreation with the master would have been nice.
I can see your reasons for making the users do this themselves if this is their goal.

>Wouldn't the simplest solution be to include the .git/ directory that represents the actual state of the container?

Yes, I am only trying to avoid the situation where I cannot work on my own branch as it replaces my changes each time I build it. 
If you prefer to have the container fetch its files during the build process, we can go that route. It would just have to be kept in mind for people who are forking in the future. 
Should we could leave in a comment about for anyone who would like to work on local changes? 
```git clone jjjchens235/bing-rewards #change to COPY .... If you want to use local files```

Killerherts commented:
This will make the container be at parity when you ```docker run -t -d``` 
It also adds a cron entry to run update.sh once a day at 00:00
This will break the container until fixes in driver.py have been applied to the master branch.

jjjchens235 commented:
Thank you! 
Can you update DockerHub with the updated image? Want to test using the final image, and too lazy to build it.

Killerherts commented:
> Thank you! Can you update DockerHub with the updated image? Want to test using the final image, and too lazy to build it.

pironic commented on compose.yaml:
```diff
@@ -0,0 +1,16 @@
+          - UMASK=022
+          - TZ=<insert to match env>
+        volumes:
+            - '<location of your config file>:/BingRewards/config'
```
`/bing-rewards/BingRewards/config`

pironic commented on compose.yaml:
```diff
@@ -0,0 +1,16 @@
+          - TZ=<insert to match env>
+        volumes:
+            - '<location of your config file>:/BingRewards/config'
+            - '<path to where you want to collect logs>:/BingRewards/logs'
```
`/bing-rewards/BingRewards/logs`

Killerherts replied:
I definitely missed handling the logs since we now are cloning during entry.sh
We could handle this by ```ln /bing-rewards/BingRewards/logs /logs```
We just cannot mount volumes inside bing-rewards anymore

pironic commented on entry.sh:
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+set -x
+service cron start
+crontab /etc/cron.d/bing.cron
```
Can I propose that we add an environment variable that allows us to disable the use of cron within the container? This is the place we can check for the existence of the env var and avoid starting the cron service entirely. Additionally when we output the log later, we just end the container if it's set. This will allow the user to schedule the container with external tools if they so chose. 

line 3-4:
```
if [[ -v NOCRON ]];
then
    echo "skipping start of cron"
else
    service cron start
    crontab /etc/cron.d/bing.cron
fi
```

line 13:
``` 
if [[ -v NOCRON ]];
then
    tail /bing-rewards/BingRewards/cronBing.log
else
    tail -f /bing-rewards/BingRewards/cronBing.log
fi

Killerherts replied:
I am not in favor of this. You can just comment out the cron entry and cron uses little memory. If you feel like stopping, it you could just exec into the container and stop it. 

If this was the direction, we wanted to take it i would not have left the container running. It would be a better option to remove tail completely and let what ever external tool your using run the whole container. If your goal is to use external tools I would fork my intention with this feature is to not need any external tools as I think that was the orginal nature of the docker request as people have run into env issues.

pironic replied:
Alright. I just wanted to query your thoughts on it... which is why i kept it out of the code-review. :-)

Killerherts commented on README.md:
```diff
@@ -36,6 +36,14 @@ actually translates to `python BingRewards.py -r -hl -d chrome`, i.e run the rem
+1. In terminal, run `docker pull killerherts/bing-rewards:latest`
+2. Set-up the config with either option 1 or 2 
+	1. Option 1, run setup.py again: `docker run -t -d --name bing-rewards killerherts/bing-rewards:latest python setup.py -e <your_email> -p <password>`  You must include your password as there will be no user prompt with -t -d
+	2. Option 2: Pass your config volume directly into the container: `docker run -t -d -v <absolute-path-to-config-directory>:/config --name bing-rewards killerherts/bing-rewards:latest`
```
> This doesn't look like it actually polls from `/config` does it? Shouldn't it be `/bing-rewards/BingRewards/config`? With current instructions it will not work.

no as the workdir is already BingRewards your already in that folder

Killerherts replied:
Because of the way we now clone git repo during the entry.sh execution during docker run the bing-rewards folder has to be empty until the repo is cloned. So my solution was just to mv the config file after the git had been cloned

pironic replied:
how about using entry.sh to make a symlink after the git clone? `ln -s /config /bing-rewards/BingRewards/config`

pironic replied:
I can see someone changing the config to add things like discord webhook info as they read the docs. In order for those changes to propagate the user just has to restart the container and the entry.sh will execute again. Instead of changes being picked up on next cron run we have to force restart the container. Just something to be aware of I guess.

Killerherts replied:
Hmm I had not really thought of that. I can see where this could get confusing if someone doesn't understand the import was not updated.  
Your solution is likely a more desirable and expected outcome I'll make the change.

Killerherts commented on entry.sh:
```diff
@@ -0,0 +1,14 @@
+pip install --no-warn-script-location -r /bing-rewards/BingRewards/requirements.txt
+cd /bing-rewards/BingRewards
+"$@"
+tail -f /dev/null
```
The work dir is already bing-rewards when container is run this is the folder your left in

pironic replied:
additional thought... not just the cronBing.log but all logs
`tail -f /bing-rewards/BingRewards/logs/*log`

Killerherts replied:
This is now added great suggestion.

pironic commented on README.md:
```diff
@@ -42,7 +42,8 @@ Here's an example of running non-default arguments
 	2. Option 2: Pass your config volume directly into the container: `docker run -t -d -v <absolute-path-to-config-directory>:/config --name bing-rewards killerherts/bing-rewards:latest`
 4. To enter the container for maintance docker exec -it bing-rewards /bin/bash 
-Note: Intially the container will be setup to run script once every 8 hours this can be modified using docker exec -it bing-rewards /bin/bash crontab -e 
+Notes: Intially the container will be setup to run script once every 8 hours this can be modified using docker exec -it bing-rewards /bin/bash crontab -e 
```
I hate to be picky but can we adjust the following for markdown readability, please:
```
Notes: Intially the container will be setup to run script once every 8 hours this can be modified using docker exec -it bing-rewards /bin/bash crontab -e
```
to
``` 
Notes: Intially the container will be setup to run script once every 8 hours this can be modified using `docker exec -it bing-rewards /bin/bash crontab -e`
```

pironic reviewed (changes requested):
Just a minor readability change for markdown. 

Thank you very much for all the changes Killerherts!

Killerherts commented:
I have changed entry.sh so it does a check for the config file and if it exists, it sets up a symbolic link like @pironic suggested 
I also added @pironic suggestions on tail command.
Added notes for people who want to mount logs to a directory on the host.

@jjjchens235  if you would like me to close the PR and fix my commit history by rebasing again I can. I force pushed before rebasing and got myself stuck with messy history since its part of the pr already.

Killerherts commented:
Moved to new PR so I could provide clean commit history. 
@pironic let me know if you see anything else.
